packages/git/src/browser/scm-container-widget.ts

```diff
@@ -0,0 +1,94 @@
+
+import { injectable, inject, multiInject, postConstruct } from 'inversify';
+import { EditorManager } from '@theia/editor/lib/browser';
+import { Repository } from '../common';
+import { GitRepositoryProvider } from './git-repository-provider';
+import { BaseWidget } from '@theia/core/lib/browser/widgets/widget';
+import { GitErrorHandler } from './git-error-handler';
+import { SingletonLayout } from '@phosphor/widgets/lib/singletonlayout';
+import { FileSystem } from '@theia/filesystem/lib/common';
+import { WidgetManager } from '@theia/core/lib/browser/widget-manager';
+import { ScmWidgetFactory } from './index';
+
+@injectable()
+export class ScmContainerWidget extends BaseWidget {
+
+    @inject(EditorManager)
+    protected readonly editorManager: EditorManager;
+
+    @inject(GitErrorHandler)
+    protected readonly gitErrorHandler: GitErrorHandler;
+
+    protected singletonLayout: SingletonLayout;
+
+    constructor(
+        @multiInject(ScmWidgetFactory) public readonly scmWidgetFactories: ScmWidgetFactory[],
+        @inject(WidgetManager) protected widgetManager: WidgetManager,
+        @inject(GitRepositoryProvider) protected readonly repositoryProvider: GitRepositoryProvider,
+        @inject(FileSystem) protected readonly fileSystem: FileSystem,
+    ) {
+
+        super();
+        this.id = 'theia-gitContainer';
+        this.addClass('theia-git');
+        this.node.tabIndex = 0;
+        this.singletonLayout = new SingletonLayout();
+        if (scmWidgetFactories.length === 1) {
+            const theOnlyWidgetId = scmWidgetFactories[0].widgetId;
+            this.widgetManager.getOrCreateWidget(theOnlyWidgetId).then(widget => {
+                this.title.label = widget.title.label;
+                this.title.caption = widget.title.caption;
+                this.title.iconClass = widget.title.iconClass;
+            });
+        } else {
+            this.title.label = 'SCM';
+            this.title.caption = 'SCM';
+            this.title.iconClass = 'fa git-tab-icon';
+        }
+        this.layout = this.singletonLayout;
+    }
+
+    @postConstruct()
+    protected init() {
+        this.toDispose.push(this.repositoryProvider.onDidChangeRepository(repository =>
+            this.initialize(repository)
+        ));
+        this.initialize(this.repositoryProvider.selectedRepository);
+    }
+
+    protected async initialize(repository: Repository | undefined): Promise<void> {
+        const widgetId = await this.getWidgetIdForRepository(repository);
+        const widget = await this.widgetManager.getOrCreateWidget(widgetId);
+        this.singletonLayout.widget = widget;
+        this.update();
+    }
+
+    protected async getWidgetIdForRepository(repository: Repository | undefined): Promise<string> {
+        if (repository) {
+            for (const factory of this.scmWidgetFactories) {
+                const fileStat = await this.fileSystem.getFileStat(repository.localUri);
```

Comment:
Do you need to await the filesystem when trying each factories?

Reply:
No, I will run these in parallel.  isUnderSourceControl is synchronous and assumes the FileStat is populated with the children.